### ISSUE TITLE
where & how ML-KNN is used ?

### ISSUE THREAD

**Issue by iTomxy** [PKU-ICST-MIPL/SSDH_TCSVT2017#3]
The README says that the **flickr** dataset is used for example, which is a **multi-label** dataset. As said in the paper, the pseudo labels should be constructed by the ML-KNN algorithm.
But as I visualize the network (*flickr25k_semirank_train_test_vggf_12b.prototxt*) and look into *semirank_loss_layer.cpp*, it seems that the pseudo labels are predicted based on the `pred`, with the id yielding the largest value denoted as `pseudo_label`, and the smallest `pseudo_label_false`. I think this may be the method used for single label dataset.
So **where & how ML-KNN is used in the code**? Or is there any tricks used here?

**Comment by PKU-ICST-MIPL** [comment]
The training data consist of labeled data and unlabeled data. And the ML-KNN is used when we judge the semantic similarity of  one psudo-label data sample and one labeled data sample, since the labeled data in flickr25k are multi-label data.

**Comment by iTomxy** [comment]
It seems that there is only one way to judge the samantic similarity (i.e., line 164 to 186 in *semirank_loss_layer.cpp*).
So, for clearity, there is **nothing different** between single-label & multi-label data in ways of pseudo label predicting and semantic similarity judging.
Am I right?

**Comment by PKU-ICST-MIPL** [comment]
The algorithm is same for predicting pseudo labels, but we apply different algorithms for semantic similarity judging.
